Added new graphics functions in the DisplayBuffer class.

nielsnl68 opened the pull request:
- added new graphic functions for use with  upcoming switchplate component
- added option to clib a part of the screen to draw inside (1)
- add color blending functions and option to set transparant color. (2)
- removed unneeded draw_pixel_at calls while doing a filedl_rectangle()

(1) clipping will not work on the clear() and fill() functions (for now ???)
(2) work in progress

# What does this implement/fix?

Add a couple of graphics function in the Displaybuffer class to be used in the upcoming Switchplate component.
extending the (filled_)rectangle function with the option to draw rounded and added new (filled_)triangle functions
Also it is now possible to set a area where in the screen can be modified (Clipping).
Also there are some color blending function that maybe need to be moved to the color.h/.cpp file later.


## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other


**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#<esphome-docs PR number goes here>

## Test Environment

- [X] ESP32
- [ ] ESP32 IDF
- [ ] ESP8266

## Example entry for `config.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
# Example config.yaml
display:
  - id: my_display
    platform: ili9341
    model: TFT 2.4
    cs_pin: 5
    dc_pin: 16
    rotation: 90

    lambda: |-
      it.fill(id(my_green));
      it.print(2, 2, id(my_font), id(my_white), TextAlign::TOP_LEFT, "Woonkamer lampen");
      //it.strftime(210, 2, id(my_font), "%d/%m %H:%M", id(the_time).now() );
      //it.rectangle(    50, 50 ,228, 150, id(my_green));
      
      it.set_clipping( 50, 50 ,228, 150);
      it.filled_rectangle(0, 0, 320, 240,  id(my_black));

      it.rectangle(3, 28, 99, 99, id(my_blue));
      it.rectangle(3,136, 99, 99, 8, id(my_red));

      it.filled_rectangle(110, 28, 99, 99, id(my_blue));
      it.filled_rectangle(110,136, 99, 99, 7,  id(my_red));

      it.triangle(       268,  28,  218, 128, 318, 128, id(my_blue));
  
      it.filled_triangle(268, 136,  218, 236, 318, 236, id(my_red));
  
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

nielsnl68 committed:
- added new grapic functions for use with
  upcoming switchplate component
- added option to clib a part of the screen to draw inside
- add color blending functions and option to set transparant color. (*)
- removed unneeded draw_pixel_at calls while doing a filedl_rectangle()

(*) work inprogress

nielsnl68 committed:
fix Trailing whitespace detected

nielsnl68 committed:
update protected properties clang issues

nielsnl68 committed:
update protected properties clang issues

nielsnl68 committed:
clang error corrections

nielsnl68 committed:
clang error corrections second run

nielsnl68 committed:
clang error corrections 3th run

nielsnl68 committed:
clang error corrections 4th run

nielsnl68 committed:
clang error corrections 5th run

nielsnl68 committed:
:(

nielsnl68 committed:
:'(

jesserockz commented on esphome/components/display/display_buffer.h:
```diff
@@ -18,6 +20,9 @@
 #include "esphome/components/qr_code/qr_code.h"
 #endif
 
+#define USE_EXTENDEDDRAW
```
Whats the point of the ifdefs if this is always defined here?

nielsnl68 replied:
Ha yes, good point i added, it just there from while i was testing. I needed a point to find the changes i made and to debug the code.
While preparing the PR i was thinking, maybe it can be for use when the added code is not wanted for the normal use outside my upcoming SwitchPlate component. 

And thanks for looking at my PR.

GregJArnold commented on esphome/components/display/display_buffer.cpp:
```diff
@@ -70,21 +81,21 @@ void HOT DisplayBuffer::draw_pixel_at(int x, int y, Color color) {
   App.feed_wdt();
 }
 void HOT DisplayBuffer::line(int x1, int y1, int x2, int y2, Color color) {
-  const int32_t dx = abs(x2 - x1), sx = x1 < x2 ? 1 : -1;
-  const int32_t dy = -abs(y2 - y1), sy = y1 < y2 ? 1 : -1;
-  int32_t err = dx + dy;
+  const int32_t delta_x = abs(x2 - x1), sx = x1 < x2 ? 1 : -1;
+  const int32_t delta_y = -abs(y2 - y1), sy = y1 < y2 ? 1 : -1;
+  int32_t err = delta_x + delta_y;
```
`dx` is a standard shorthand for `delta_x`, so not only is the variable name longer, but the changes increase the diff set, which makes it harder to see what's changed.  You should revert these changes, here and below.

GregJArnold commented on esphome/components/display/display_buffer.cpp:
```diff
@@ -392,6 +402,484 @@ void DisplayBuffer::strftime(int x, int y, Font *font, const char *format, time:
+  return color_new;
+}
+
+bool DisplayBuffer::is_color_equal(Color a, Color b) { return a.r == b.r && a.g == b.g && a.b == b.b; }
```
`Color` already has a `=` operator to compare two colors.  Can you not use that?

nielsnl68 replied:
I agree, this function could be removed indeed.

GregJArnold commented on esphome/components/display/display_buffer.cpp:
```diff
@@ -392,6 +402,484 @@ void DisplayBuffer::strftime(int x, int y, Font *font, const char *format, time:
+  }
+}
+
+void DisplayBuffer::quad(Point *points, Color color) {
+  int16_t x0, y0, x1, y1;
+
+  x0 = points[0].x;
+  y0 = points[0].y;
+  x1 = points[1].x;
+  y1 = points[1].y;
+  this->line(x0, y0, x1, y1, color);
+
+  x0 = points[1].x;
+  y0 = points[1].y;
+  x1 = points[2].x;
+  y1 = points[2].y;
+  this->line(x0, y0, x1, y1, color);
+
+  x0 = points[2].x;
+  y0 = points[2].y;
+  x1 = points[3].x;
+  y1 = points[3].y;
+  this->line(x0, y0, x1, y1, color);
+
+  x0 = points[3].x;
+  y0 = points[3].y;
+  x1 = points[0].x;
+  y1 = points[0].y;
+  this->line(x0, y0, x1, y1, color);
+}
```
Given that you've created a `Point`, I'd suggest creating additional function signatures that take a `Point` instead of an (x, y) pair, then this can be greatly simplified:
```suggestion
void DisplayBuffer::quad(Point *points, Color color) {
  this->line(points[0], points[1], color);
  this->line(points[1], points[2], color);
  this->line(points[2], points[3], color);
  this->line(points[3], points[0], color);
}
```

nielsnl68 replied:
can you explain more, could you tell how that new function should be look like?

GregJArnold replied:
You'd create a function something like this:
```cpp
void DisplayBuffer::line(Point &pt1, Point &pt2, Color color,) {
  this->line(pt1.x, pt1.y, pt2.x, pt2.y, color);
}

nielsnl68 replied:
okey, i got you, I will add it when i see more then 5 thumbs-up otherwise i go the other way around and remove the point version in favor of:
```cpp
void qaud(int16_t x0, int16_t y0, int16_t x1, int16_t y1, int16_t x2, int16_t y2, int16_t x3, int16_t y3, Color color = COLOR_ON) { 
  ...
}
```

nielsnl68 replied:
I have removed the point structure from the PR changes,

GregJArnold commented on esphome/components/display/display_buffer.cpp:
```diff
@@ -392,6 +402,484 @@ void DisplayBuffer::strftime(int x, int y, Font *font, const char *format, time:
+// two filled triangles sharing one side. We have to be careful
+// about the triangle fill routine (ie. using rounding) so that
+// we can avoid leaving a thin seam between the two triangles.
+void DisplayBuffer::filled_quad(Point *points, Color color) {
+  int16_t x0, y0, x1, y1, x2, y2;
+
+  // Break down quadrilateral into two triangles
+  x0 = points[0].x;
+  y0 = points[0].y;
+  x1 = points[1].x;
+  y1 = points[1].y;
+  x2 = points[2].x;
+  y2 = points[2].y;
+  this->filled_triangle(x0, y0, x1, y1, x2, y2, color);
+
+  x0 = points[2].x;
+  y0 = points[2].y;
+  x1 = points[0].x;
+  y1 = points[0].y;
+  x2 = points[3].x;
+  y2 = points[3].y;
+  this->filled_triangle(x0, y0, x1, y1, x2, y2, color);
+}
```
Same as above:
```suggestion
void DisplayBuffer::filled_quad(Point *points, Color color) {
  // Break down quadrilateral into two triangles
  this->filled_triangle(points[0], points[1], points[2], color);
  this->filled_triangle(points[2], points[0], points[3], color);
}
```

nielsnl68 replied:
I did something similar as this, by removing the `Point` structure.

GregJArnold reviewed (changes requested):
Looks good; just a few suggestions.

nielsnl68 committed:
Add Gradient function to the color class.

nielsnl68 committed:
fix spaceing

nielsnl68 committed:
remove more spaces

nielsnl68 committed:
Merge branch 'esphome:dev' into dev

nielsnl68 committed:
- removed the Point structure and replaced it
  with comon params
- removed and moved the color fuctions to the
  color object. see PR#3687
- the clipping fuction is  now stackable,
  call `set_clipping()` to stack
  call 'clear_clipping()` to restore to prev stack

nielsnl68 committed:
Merge branch 'dev' of https://github.com/nielsnl68/esphome into dev

nielsnl68 committed:
fixed issues reported by clang reporter

nielsnl68 committed:
fiximg the new clang issues

nielsnl68 committed:
- fixed more clang issues
